include/mbedtls/config_adjust_legacy_from_psa.h

```diff
@@ -443,40 +443,40 @@
 #if defined(PSA_WANT_DH_RFC7919_2048)
 #if !defined(MBEDTLS_PSA_ACCEL_DH_RFC7919_2048)
 #define MBEDTLS_PSA_BUILTIN_DH_RFC7919_2048 1
-#define MBEDTLS_DHM_C
+#define MBEDTLS_PSA_DH_ACCEL_INCOMPLETE_PARAMS
 #endif /* !MBEDTLS_PSA_BUILTIN_DH_RFC7919_2048 */
 #endif /* PSA_WANT_DH_RFC7919_2048 */
 
 #if defined(PSA_WANT_DH_RFC7919_3072)
 #if !defined(MBEDTLS_PSA_ACCEL_DH_RFC7919_3072)
 #define MBEDTLS_PSA_BUILTIN_DH_RFC7919_3072 1
-#define MBEDTLS_DHM_C
+#define MBEDTLS_PSA_DH_ACCEL_INCOMPLETE_PARAMS
 #endif /* !MBEDTLS_PSA_BUILTIN_DH_RFC7919_3072 */
 #endif /* PSA_WANT_DH_RFC7919_3072 */
 
 #if defined(PSA_WANT_DH_RFC7919_4096)
 #if !defined(MBEDTLS_PSA_ACCEL_DH_RFC7919_4096)
 #define MBEDTLS_PSA_BUILTIN_DH_RFC7919_4096 1
-#define MBEDTLS_DHM_C
+#define MBEDTLS_PSA_DH_ACCEL_INCOMPLETE_PARAMS
 #endif /* !MBEDTLS_PSA_BUILTIN_DH_RFC7919_4096 */
 #endif /* PSA_WANT_DH_RFC7919_4096 */
 
 #if defined(PSA_WANT_DH_RFC7919_6144)
 #if !defined(MBEDTLS_PSA_ACCEL_DH_RFC7919_6144)
 #define MBEDTLS_PSA_BUILTIN_DH_RFC7919_6144 1
-#define MBEDTLS_DHM_C
+#define MBEDTLS_PSA_DH_ACCEL_INCOMPLETE_PARAMS
 #endif /* !MBEDTLS_PSA_BUILTIN_DH_RFC7919_6144 */
 #endif /* PSA_WANT_DH_RFC7919_6144 */
 
 #if defined(PSA_WANT_DH_RFC7919_8192)
 #if !defined(MBEDTLS_PSA_ACCEL_DH_RFC7919_8192)
 #define MBEDTLS_PSA_BUILTIN_DH_RFC7919_8192 1
-#define MBEDTLS_DHM_C
+#define MBEDTLS_PSA_DH_ACCEL_INCOMPLETE_PARAMS
 #endif /* !MBEDTLS_PSA_BUILTIN_DH_RFC7919_8192 */
 #endif /* PSA_WANT_DH_RFC7919_8192 */
 
 #if defined(PSA_WANT_ALG_FFDH)
-#if !defined(MBEDTLS_PSA_ACCEL_ALG_FFDH)
+#if !defined(MBEDTLS_PSA_ACCEL_ALG_FFDH) || defined(MBEDTLS_PSA_DH_ACCEL_INCOMPLETE_PARAMS)
```

Comment:
This is more for @mpg. What is the intended behavour?

Why not `!defined(MBEDTLS_PSA_ACCEL_ALG_FFDH) && defined(MBEDTLS_PSA_DH_ACCEL_INCOMPLETE_PARAMS)`

I am thinking of edge cases, where we have `BEDTLS_PSA_ACCEL_ALG_FFDH` and "PSA_WANT_DH_RFC7919_8192" for example?

Reply:
I try to answer here, but @mpg will surely have a more complete answer :)

The idea is that enabling builtin support for any of the `PSA_WANT_DH_RFC7919_xxx` domain parameters makes little sense if there is not algorithm using it (i.e. `MBEDTLS_PSA_BUILTIN_ALG_FFDH`) therefore we need to enable that support manually. If the end user does not want this kind of support, then he/she will need to:
- accelerate both `ALG_FFDH` and all the required domain parameters `DH_RFC7919_xxx`
- disable both (alg + domain params)

Reply:
For things that depend on multiple `PSA_WANT` options (for example AES-GCM needs `PSA_WANT_KEY_TYPE_AES` and `PSA_WANT_ALG_GCM`) the principle is that the thing is available as soon as all the `PSA_WANT` are enabled. (This means you can't enable AES-CBC and Aria-GCM without also getting AES-GCM and Aria-CBC. That is, the set of things enabled is the full cross-product of the components.)

Similarly, a driver supports AES-GCM only if it defines both `ACCEL_KEY_TYPE_AES` and `ACCEL_ALG_GCM`.

Each mechanism needs to be built-in if it's not supported by a driver, that is if not all components have an accel macro - said otherwise, if any component is missing an accel macro.

Sometimes, a mechanism being built-in means all its components are built-in (for example, if you want ECDH with secp521r1 but your driver doesn't support that, you'll need both `ECDH_C` _and_ `ECP_DP_SECP521R1`, even if your driver supports ECDH with other curves, or that curve with other algs). Sometimes, a built-in module can take advantage of a partial driver: for example if your driver support SHA-256 but not HMAC-SHA-256, you only need HMAC to be built-in (and `MD_C`, which implements built-in HMAC, will call the driver for SHA-256); or thanks to Valerio's recent work, if you have a driver that supports AES but not GCM, you can still only enable `GCM_C` without `AES_C` and still get AES-GCM.

That does not apply to (FF)DH where if anything is not accelerated, then you need everything to be built-in.

Now, while writing this I realise I forgot something about (FF)DH: there are key types involved as well. So it's actually really similar to [the situation with ECC](https://github.com/Mbed-TLS/mbedtls/blob/development/include/mbedtls/config_adjust_legacy_from_psa.h#L34) - except (FF)DH is somewhat simpler as there's only one algorithm (hence no question about which curve / parameters matter for which alg).

So, I think we should add some check for whether the driver support DH key types or not. As this point, I'd recommend structuring the (FF)DH section in the same way as the ECC section just above for uniformity. That is, two passes: one to check if everything is complete or not, then one to enable stuff based on that.